File: Threema/BlobUploader.m
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2015-2020 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

#import "BlobUploader.h"
#import "ActivityIndicatorProxy.h"
#import "Contact.h"
#import "BlobUtil.h"
#import "BaseMessage.h"
#import "ThreemaError.h"
#import "SSLCAHelper.h"
#import "NSString+Hex.h"
#import "BlobUploadDelegate.h"

#ifdef DEBUG
  static const DDLogLevel ddLogLevel = DDLogLevelVerbose;
#else
  static const DDLogLevel ddLogLevel = DDLogLevelWarning;
#endif
@interface BlobUploader ()

@property NSMutableData *receivedData;
@property NSURLConnection *uploadConnection;

@property id<BlobUploadDelegate> blobUploadDelegate;

@end

@implementation BlobUploader

- (void)startUploadFor:(id<BlobUploadDelegate>)messageProxy {
    _blobUploadDelegate = messageProxy;
    
    [ActivityIndicatorProxy startActivity];
    [_blobUploadDelegate uploadProgress: [NSNumber numberWithFloat:0]];
    
    NSURL *blobUploadUrl = [BlobUtil urlForBlobUpload];
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:blobUploadUrl cachePolicy:NSURLRequestReloadIgnoringLocalAndRemoteCacheData timeoutInterval:kBlobUploadTimeout];
    request.HTTPMethod = @"POST";
    
    NSString *boundary = @"---------------------------Boundary_Line";
    NSString *contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@",boundary];
    [request addValue:contentType forHTTPHeaderField: @"Content-Type"];
    
    NSUInteger dataLength = [self totalDataLength];
    NSMutableData *body = [NSMutableData dataWithCapacity: dataLength];
    [body appendData:[[NSString stringWithFormat:@"--%@\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];
    [body appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"blob\"; filename=\"blob.bin\"\r\n"] dataUsingEncoding:NSUTF8StringEncoding]];
    [body appendData:[@"Content-Type: application/octet-stream\r\n\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
    [body appendData:_data];
    _data = nil;    // release memory now
    
    if (_thumbnailData) {
        [body appendData:[[NSString stringWithFormat:@"\r\n--%@\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];

        [body appendData:[[NSString stringWithFormat:@"Content-Disposition: form-data; name=\"blob2\"; filename=\"blob2.bin\"\r\n"] dataUsingEncoding:NSUTF8StringEncoding]];
        [body appendData:[@"Content-Type: application/octet-stream\r\n\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
        [body appendData:_thumbnailData];
        _thumbnailData = nil;    // release memory now
        [body appendData:[[NSString stringWithFormat:@"\r\n--%@--\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];
    } else {
        [body appendData:[[NSString stringWithFormat:@"\r\n--%@--\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];
    }
    
    [request setHTTPBody:body];
    [request addValue:[NSString stringWithFormat:@"%lu", (unsigned long)[body length]] forHTTPHeaderField:@"Content-Length"];
    
    _receivedData = [NSMutableData data];
    
    _uploadConnection = [[NSURLConnection alloc] initWithRequest:request delegate:self startImmediately:NO];
    [_uploadConnection scheduleInRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
    [_uploadConnection start];
}

- (NSUInteger)totalDataLength {
    NSUInteger length = _data.length + 1024;
    if (_thumbnailData) {
        length += _thumbnailData.length;
    }
    
    return length;
}


#pragma mark - URL connection delegate

- (void)connection:(NSURLConnection *)connection didSendBodyData:(NSInteger)bytesWritten totalBytesWritten:(NSInteger)totalBytesWritten totalBytesExpectedToWrite:(NSInteger)totalBytesExpectedToWrite {
    DDLogVerbose(@"totalBytesWritten: %ld, totalBytesExpectedToWrite: %ld", (long)totalBytesWritten, (long)totalBytesExpectedToWrite);
    
    if ([_blobUploadDelegate uploadShouldCancel]) {
        DDLogWarn(@"Upload cancelled");
        
        [connection cancel];
        
        [_blobUploadDelegate uploadDidCancel];
        return;
    }
    
    [_blobUploadDelegate uploadProgress: [NSNumber numberWithFloat:((float)totalBytesWritten / (float)totalBytesExpectedToWrite)]];
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
    DDLogVerbose(@"Request succeeded - received %lu bytes", (unsigned long)[_receivedData length]);
    [ActivityIndicatorProxy stopActivity];
    
    NSString *blobIdHex = [[NSString alloc] initWithData:_receivedData encoding:NSASCIIStringEncoding];
    NSData *blobId = [blobIdHex decodeHex];
    
    if (blobId.length == kBlobIdLen) {
        DDLogVerbose(@"Blob ID: %@", blobId);
        [_blobUploadDelegate uploadSucceededWithBlobIds:@[blobId]];
    } else if (blobId.length == (2*kBlobIdLen)) {
        NSData *blobId1 = [NSData dataWithBytes:blobId.bytes length:kBlobIdLen];
        NSData *blobId2 = [NSData dataWithBytes:(blobId.bytes + kBlobIdLen) length:kBlobIdLen];
        
        DDLogVerbose(@"Blob ID: %@ / %@", blobId1, blobId2);
        [_blobUploadDelegate uploadSucceededWithBlobIds:@[blobId1, blobId2]];
    } else {
        DDLogError(@"Got invalid blob ID from server: %@", blobId);
        [self connection:connection didFailWithError:[ThreemaError threemaError:@"Got invalid blob ID from server"]];
    }
    
    _uploadConnection = nil;    // release memory for body etc. now
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
    
    DDLogError(@"Connection failed - error %@ %@",
               [error localizedDescription],
               [[error userInfo] objectForKey:NSURLErrorFailingURLStringErrorKey]);
    
    [ActivityIndicatorProxy stopActivity];
    
    if ([_blobUploadDelegate uploadShouldCancel]) {
        DDLogWarn(@"Upload cancelled");
        
        [_blobUploadDelegate uploadDidCancel];
        return;
    }
    
    [_blobUploadDelegate uploadFailed];
    
    _uploadConnection = nil;    // release memory for body etc. now
}

/* this method ensures that HTTP errors get treated like connection failures etc. as well */
- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
    [_receivedData setLength:0];
    
    NSInteger statusCode = [((NSHTTPURLResponse *)response) statusCode];
    if (statusCode >= 400)
    {
        [connection cancel];  // stop connecting; no more delegate messages
        NSDictionary *errorInfo = [NSDictionary dictionaryWithObject:[NSString stringWithFormat:
                                                                      NSLocalizedString(@"Server returned status code %d",@""),
                                                                      statusCode]
                                                              forKey:NSLocalizedDescriptionKey];
        NSError *statusError = [NSError errorWithDomain:NSURLErrorDomain
                                                   code:statusCode
                                               userInfo:errorInfo];
        [self connection:connection didFailWithError:statusError];
    }
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
    [_receivedData appendData:data];
}

- (BOOL)connection:(NSURLConnection *)connection canAuthenticateAgainstProtectionSpace:(NSURLProtectionSpace *)protectionSpace {
    return [SSLCAHelper connection:connection canAuthenticateAgainstProtectionSpace:protectionSpace];
}

- (void)connection:(NSURLConnection *)connection didReceiveAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge {
    [SSLCAHelper connection:connection didReceiveAuthenticationChallenge:challenge];
}

- (NSCachedURLResponse *)connection:(NSURLConnection *)connection willCacheResponse:(NSCachedURLResponse *)cachedResponse {
    return nil;
}

@end

File: ThreemaFramework/Core/BlobUploader.swift
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2022-2025 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import CocoaLumberjackSwift
import Foundation
import PromiseKit

protocol BlobUploaderProtocol {
    func upload(
        data: Data,
        origin: BlobOrigin,
        setPersistParam: Bool
    ) -> Promise<Data>
}

class BlobUploader: NSObject, BlobUploaderProtocol {
    
    enum BlobUploaderError: Error {
        case uploadFailed(message: String)
        case invalidUploadURL
        case idGenerationFailed
    }
    
    private let blobURL: BlobURL
    private let queue: DispatchQueue
    private var urlSessionManager: URLSessionManager
    
    private var progressObservers = [NSManagedObjectID: NSKeyValueObservation]()
    private var activeTasks = [NSManagedObjectID: URLSessionTask]()
    
    // MARK: - Lifecycle
    
    init(
        blobURL: BlobURL,
        queue: DispatchQueue = DispatchQueue.main,
        sessionManager: URLSessionManager = .shared
    ) {
        self.blobURL = blobURL
        self.queue = queue
        self.urlSessionManager = sessionManager
        super.init()
    }
    
    // MARK: - Upload
    
    /// Uploads given data and provides progress updates to its delegate
    /// - Parameters:
    ///   - blobData: Data to be uploaded
    ///   - origin: Origin of the Blob
    ///   - objectID: ObjectID used to track progress
    ///   - delegate: Delegate to send progress updates to
    /// - Returns: Received data from server
    @discardableResult
    func upload(
        blobData: Data,
        origin: BlobOrigin,
        setPersistParam: Bool,
        objectID: NSManagedObjectID,
        delegate: BlobManagerDelegate? = nil
    ) async throws -> Data {
        try await withCheckedThrowingContinuation { continuation in
            blobURL.upload(
                origin: origin,
                setPersistParam: setPersistParam,
                completionHandler: { uploadURL, authorization, error in
                
                    if let error {
                        continuation.resume(throwing: error)
                        return
                    }
                
                    guard let uploadURL else {
                        continuation.resume(throwing: BlobUploaderError.invalidUploadURL)
                        return
                    }
                
                    let task = self
                        .upload(for: uploadURL, and: blobData, authorization: authorization) { data, error in
                        
                            // Upload was completed
                            if let error {
                                continuation.resume(throwing: error)
                            }
                            else if let data {
                                // We got data, so we create the ID for it
                                guard let blobIDHex = String(bytes: data, encoding: .ascii),
                                      let blobID = BytesUtility.toBytes(hexString: blobIDHex) else {
                                    continuation.resume(throwing: BlobUploaderError.idGenerationFailed)
                                    return
                                }
                            
                                let id = Data(blobID)
                                continuation.resume(returning: id)
                            }
                            else {
                                continuation.resume(
                                    throwing: BlobUploaderError
                                        .uploadFailed(message: "[BlobUploader] Upload unsuccessful.")
                                )
                            }
                        
                            // Remove completed Task
                            self.progressObservers.removeValue(forKey: objectID)
                            self.activeTasks.removeValue(forKey: objectID)
                        }
                
                    // Add created task to keep track
                    self.activeTasks[objectID] = task
                    let observer = task.progress.observe(\.fractionCompleted) { progress, _ in
                        Task {
                            await delegate?.updateProgress(for: objectID, didUpdate: progress)
                        }
                    }
                    self.progressObservers[objectID] = observer
                }
            )
        }
    }

    func upload(
        data: Data,
        origin: BlobOrigin,
        setPersistParam: Bool
    ) -> Promise<Data> {
        Promise { seal in
            blobURL.upload(
                origin: .local,
                setPersistParam: setPersistParam,
                completionHandler: { uploadURL, authorization, error in
                    if let error {
                        seal.reject(error)
                        return
                    }

                    guard let uploadURL else {
                        seal.reject(BlobUploaderError.invalidUploadURL)
                        return
                    }

                    self.upload(for: uploadURL, and: data, authorization: authorization) { data, error in
                        if let error {
                            seal.reject(error)
                            return
                        }

                        guard let data,
                              let idHex = String(bytes: data, encoding: .ascii),
                              let id = BytesUtility.toBytes(hexString: idHex) else {
                            seal.reject(BlobUploaderError.uploadFailed(message: "[BlobUploader] Upload unsuccessful."))
                            return
                        }

                        seal.fulfill(Data(id))
                    }
                }
            )
        }
    }

    @discardableResult
    private func upload(
        for url: URL,
        and data: Data,
        authorization: String?,
        completion: @escaping (Data?, Error?) -> Void
    ) -> URLSessionDataTask {
        
        let client = HTTPClient(authorization: authorization, sessionManager: urlSessionManager)
        let task = client
            .uploadData(
                url: url,
                data: createUploadData(with: data),
                contentType: .multiPart
            ) { data, response, error in
            
                var uploadError: Error?
            
                if let error {
                    uploadError = error
                }
                else {
                    if response is HTTPURLResponse {
                        if let response = response as? HTTPURLResponse,
                           !(200...299).contains(response.statusCode) {
                            uploadError = BlobUploaderError
                                .uploadFailed(
                                    message: "[BlobUploader] Upload failed with response code: \(response.statusCode)."
                                )
                        }
                    }
                    else {
                        uploadError = BlobUploaderError.uploadFailed(message: "[BlobUploader] Response is missing")
                    }
                }
                self.queue.async {
                    completion(data, uploadError)
                }
            }
        return task
    }
    
    /// Cancels a upload for given objectID if it exists
    /// - Parameters:
    ///   - objectID: ObjectID of blob
    public func cancelUpload(for objectID: NSManagedObjectID) {
        guard let task = activeTasks[objectID] else {
            return
        }
        
        task.cancel()
        activeTasks.removeValue(forKey: objectID)
        progressObservers.removeValue(forKey: objectID)
    }
    
    // MARK: - Helpers
    
    private func createUploadData(with blobData: Data) -> Data {
        let boundary = "---------------------------Boundary_Line"
        
        var data = Data()
        data.append(contentsOf: "--\(boundary)\r\n".utf8)
        data.append(contentsOf: "Content-Disposition: form-data; name=\"blob\"; filename=\"blob.bin\"\r\n".utf8)
        data.append(contentsOf: "Content-Type: application/octet-stream\r\n\r\n".utf8)
        data.append(blobData)
        data.append(contentsOf: "\r\n--\(boundary)--\r\n".utf8)
        
        return data
    }
}
